In [29]:
import torch
import numpy as np
import torch.nn as nn

from torchvision import transforms
from torchvision.datasets import ImageFolder
from torch.utils.data import DataLoader


### Creating Tensor

In [2]:
# using torch.tensor
t = torch.Tensor([[1,2,3], [3,4,5]])
t

tensor([[1., 2., 3.],
        [3., 4., 5.]])

In [3]:
# using torch.randn
t = torch.randn(3,5)
t

tensor([[-2.5975,  0.2973, -0.7316, -0.5976,  0.7884],
        [-1.7344, -0.4889, -0.2703, -0.0820, -0.2320],
        [-1.3007, -1.0921,  0.9187,  0.0962, -0.7346]])

In [4]:
# using torch.ones
t = torch.ones(3,5)
t

tensor([[1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.],
        [1., 1., 1., 1., 1.]])

In [5]:
# using torch.zeros
t = torch.zeros(3,5)
t

tensor([[0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.],
        [0., 0., 0., 0., 0.]])

In [6]:
# using torch.randn
t = torch.randint(low=0, high=5, size=(3,5))
t

tensor([[3, 4, 4, 0, 3],
        [3, 0, 3, 2, 2],
        [4, 0, 2, 3, 0]])

In [7]:
# tensor from numpy array
a = np.array([[1,2,3], [4,5,6]])
t = torch.from_numpy(a)
t

tensor([[1, 2, 3],
        [4, 5, 6]])

In [8]:
# numpy from tensor
a = t.numpy()
a

array([[1, 2, 3],
       [4, 5, 6]])

### Tensor Operation

In [9]:
A = torch.randn(3,4)
W = torch.randn(4,2)

#multiply matrix A, W
t = A.mm(W)
t

tensor([[-0.4942,  1.1303],
        [-0.0312, -2.3750],
        [ 0.2509, -0.9964]])

In [10]:
# transpose tensor
t = t.t()
t

tensor([[-0.4942, -0.0312,  0.2509],
        [ 1.1303, -2.3750, -0.9964]])

In [11]:
# square each element of t
t = t**2
t

tensor([[2.4420e-01, 9.7348e-04, 6.2955e-02],
        [1.2775e+00, 5.6407e+00, 9.9280e-01]])

In [12]:
# size of the t
t.size()

torch.Size([2, 3])

### The nn.Module

In [13]:
# How to define nerural network in PyTorch
class myNueralNet(nn.Module):
    def __init__(self):
        super().__init__()
        # Define all the layers here
        self.lin1 = nn.Linear(784, 30)
        self.lin2 = nn.Linear(30, 30)
        
    def forward(self, x):
        x = self.lin1(x)
        x = self.lin2(x)
        return x
    

But the thing to note is that we can define any sort of calculation while defining the forward pass, and that makes PyTorch highly customizable for research purposes.

In [14]:
class myCrazyNeuralNet(nn.Module):
    def __init__(self):
        super().__init__()
        #define all the layers
        self.lin1 = nn.Linear(784, 30)
        self.lin2 = nn.Linear(30, 784)
        self.lin3 = nn.Linear(30, 10)
        
    def forward(self, x):
        # connection of the layers
        x_lin1 = self.lin1(x)
        x_lin2 = x + self.lin2(x_lin1)
        x_lin2 = self.lin1(x_lin2)
        x = self.lin3(x_lin2)
        

In [15]:
x = torch.randn((100, 784))
model = myCrazyNeuralNet()
#model(x).size()

### Custom Linear Layer

In [16]:
# Defining custom linear layers
class myCustomLinearLayer(nn.Module):
    def __init__(self, in_size, out_size):
        super().__init__()
        self.weights = nn.Parameter(torch.randn(in_size, out_size))
        self.bias = nn.Parameter(torch.zeros(out_size))
    
    # connect custom layer    
    def forward(self, x):
        return x.mm(self.weights) + self.bias
        

In [19]:
class myCustomNeuralnet(nn.Module):
    def __init__(self):
        super().__init__()
        # Calling custom layer
        self.lin1 = myCustomLinearLayer(784, 10)
        
    def forward(self, x):
        # connect the layer
        x = self.lin1(x)
        return x
    
x = torch.randn(100, 784)
model = myCustomNeuralnet()
model(x).size()
    

torch.Size([100, 10])

### Conv2D

In [22]:
conv_layer = nn.Conv2d(in_channels=3, out_channels=64, kernel_size=(3,3), stride=1, padding=1)

In [23]:
x = torch.randn((100, 3, 24,24))
conv_layer(x).size()

torch.Size([100, 64, 24, 24])

### Dataset and DataLoader

In [28]:
train_path = "/home/hasan/DATA SET/Indoor CVPR/train_image"

t = transforms.Compose([
                    transforms.Resize(size=256),
                    transforms.CenterCrop(size=224),
                    transforms.ToTensor()
                    ])

train_dataset = ImageFolder(root=train_path, transform=t)

print("Total number of image in the dataset :", len(train_dataset))
print("Example image and label :", train_dataset[2])

Total number of image in the dataset : 5360
Example image and label : (tensor([[[0.5137, 0.5098, 0.5098,  ..., 0.8353, 0.8353, 0.8314],
         [0.5176, 0.5176, 0.5137,  ..., 0.8549, 0.8314, 0.8235],
         [0.5216, 0.5176, 0.5176,  ..., 0.8706, 0.8392, 0.8314],
         ...,
         [0.6157, 0.6196, 0.6118,  ..., 0.5647, 0.5725, 0.5569],
         [0.6157, 0.6235, 0.6275,  ..., 0.5725, 0.5804, 0.5686],
         [0.6118, 0.6353, 0.6392,  ..., 0.5725, 0.5765, 0.5686]],

        [[0.4392, 0.4353, 0.4392,  ..., 0.6510, 0.6471, 0.6353],
         [0.4471, 0.4471, 0.4431,  ..., 0.6863, 0.6549, 0.6431],
         [0.4510, 0.4471, 0.4431,  ..., 0.7216, 0.6863, 0.6627],
         ...,
         [0.5059, 0.5059, 0.4824,  ..., 0.4275, 0.4353, 0.4196],
         [0.5059, 0.5137, 0.5098,  ..., 0.4353, 0.4431, 0.4314],
         [0.4941, 0.5176, 0.5137,  ..., 0.4392, 0.4392, 0.4314]],

        [[0.3098, 0.3059, 0.3059,  ..., 0.5843, 0.5647, 0.5412],
         [0.3098, 0.3098, 0.2980,  ..., 0.6353, 0.58

This dataset has 5360 images, and we can get an image and its label using an index. Now we can pass images one by one to any image neural network using a for loop:

In [ ]:
for i in range(0, len(train_dataset)):
    image, label = train_dataset[i]
    pred = model(image)

But that is not optimal. We want to do batching and using DataLoader

In [31]:
train_dataloader = DataLoader(train_dataset, batch_size=64, shuffle=True, num_workers=10)

In [33]:
for image_batch, label_batch in train_dataloader:
    print(image_batch.size(), label_batch.size())
    break

torch.Size([64, 3, 224, 224]) torch.Size([64])


### The whole process of dataloading

In [ ]:
train_path = "/home/hasan/DATA SET/Indoor CVPR/train_image"

t = transforms.Compose([
                    transforms.Resize(size=256),
                    transforms.CenterCrop(size=224),
                    transforms.ToTensor()
                    ])

train_dataset = ImageFolder(root=train_path, transform=t)

train_dataloader = DataLoader(train_dataset, batch_size=64, shuffle=True, num_workers=10)

for image_batch, label_batch in train_dataloader:
    pred = myImageNeuralNet(image_batch)
    

the main power of Pytorch comes with its immense customization. We can also create our own custom datasets if the datasets provided by PyTorch don’t fit our use case.

### Understanding Custom Datasets

In [39]:
from glob import glob
from PIL import Image
from torch.utils.data import Dataset


In [40]:
class customImageFolderDataset(Dataset):
    def __init__(self, root, transform=None):
        
        self.image_paths = glob(f"{root}/*/*")
        
        self.labels = [x.split("/")[-2] for x in self.image_paths]
        
        self.label_to_idx = {x:i for i,x in enumerate(set(self.labels))}
        self.transform = transform
        
    def __len__(self):
        # return length of dataset
        return len(self.image_paths)
    
    def __getitem__(self, idx):
        # open and send one image and label
        img_name = self.image_paths[idx]
        label = self.labels[idx]
        image = Image.open(img_name)
        
        if self.transform:
            image = self.transform(image)
        return image, self.label_to_idx[label]
         

In [43]:
train_path = "/home/hasan/DATA SET/Indoor CVPR/train_image"

t = transforms.Compose([
                    transforms.Resize(size=256),
                    transforms.CenterCrop(size=224),
                    transforms.ToTensor()
                    ])

train_dataset = customImageFolderDataset(root=train_path, transform=t)
train_dataloader = DataLoader(train_dataset,batch_size = 64, shuffle=True, num_workers=10)

for image_batch, label_batch in train_dataloader:
    pred = myImageNeuralNet(image_batch)


### Understanding Custom DataLoaders

### Training a Neural Network

We know how to create a neural network using nn.Module. But how to train it? Any neural network that has to be trained will have a training loop that will look something similar to below:

In [ ]:
num_epochs = 5

for epoch in range(num_epochs):
    #set model to train mode
    model.train()
    for x_batch, y_batch in train_dataloader:
        
        #clear gradients
        optimizer.zero_grad()
        
        #forward pass - predicted output
        pred = model(x_batch)
        
        # find loss and backpropagation of gradients
        loss= loss_criterion(pred, y_batch)
        loss.backward()
        
        #update the parameters
        optimizer.step()
        
    model.eval()
    for x_batch, y_batch in valid_dataloader:
        pred = model(x_batch)
        val_loss = loss_criterion(pred, y_batch)
    

### Loss Function

Pytorch provides us with a variety of loss functions for our most common tasks, like Classification and Regression. Some most used examples are 

1. nn.CrossEntropyLoss , 
2. nn.NLLLoss , 
3. nn.KLDivLoss and 
4. nn.MSELoss

So, we can try to use this Loss function for a simple classification network. Please note the LogSoftmax layer after the final linear layer. If you don't want to use this LogSoftmax layer, you could have just used nn.CrossEntropyLoss

In [44]:
class myClassificationNet(nn.Module):
    def __init__(self):
        super().__init__()
        # Define all Layers Here
        self.lin = nn.Linear(784, 10)
        self.logsoftmax = nn.LogSoftmax(dim=1)
    def forward(self, x):
        # Connect the layer Outputs here to define the forward pass
        x = self.lin(x)
        x = self.logsoftmax(x)
        return x

In [45]:
# some random input:
X = torch.randn(100,784)
y = torch.randint(low = 0,high = 10,size = (100,))

In [47]:
y

tensor([5, 0, 7, 7, 0, 7, 5, 3, 5, 5, 0, 9, 7, 6, 1, 9, 9, 0, 8, 5, 3, 3, 4, 9,
        9, 8, 7, 1, 3, 0, 2, 7, 7, 5, 5, 7, 3, 1, 1, 7, 0, 7, 3, 2, 3, 2, 6, 6,
        6, 6, 0, 4, 4, 7, 6, 4, 5, 5, 8, 7, 5, 3, 4, 1, 6, 1, 4, 8, 6, 0, 0, 5,
        1, 1, 3, 8, 1, 4, 0, 7, 0, 0, 6, 1, 4, 6, 0, 2, 6, 0, 9, 6, 9, 7, 7, 4,
        1, 6, 8, 7])

In [48]:
# And pass it through the model to get predictions:
model = myClassificationNet()
preds = model(X)

In [49]:
#We can now get the loss as:
criterion = nn.NLLLoss()
loss = criterion(preds,y)
loss

tensor(2.4575, grad_fn=<NllLossBackward>)

### Custom Loss Function

Defining your custom loss functions is again a piece of cake, and you should be okay as long as you use tensor operations in your loss function. For example, here is the customMseLoss

In [53]:
def customMseLoss(output, target):
    loss = torch.mean((output - target)**2)
    return loss

You can use this custom loss just like before. But note that we don’t instantiate the loss using criterion this time as we have defined it as a function.

In [ ]:
output = model(X)
loss = customMseLoss(output, target)
loss.backward()

If we wanted, we could have also written it as a class using nn.Module , and then we would have been able to use it as an object. Here is an NLLLoss custom example:

In [ ]:
class CustomNLLLoss(nn.Module):
    def __init__(self):
        super().__init__()
    def forward(self, x, y):
        # x should be output from LogSoftmax Layer 
        log_prob = -1.0 * x
        # Get log_prob based on y class_index as loss=-mean(ylogp)
        loss = log_prob.gather(1, y.unsqueeze(1))
        loss = loss.mean()
        return loss
criterion = CustomNLLLoss()
loss = criterion(preds,y)

### Optimizers

Once we get gradients using the loss.backward() call, we need to take an optimizer step to change the weights in the whole network. Pytorch provides a variety of different ready to use optimizers using the torch.optim module. For example: 
1. torch.optim.Adadelta , 
2. torch.optim.Adagrad , 
3. torch.optim.RMSprop and the most widely used 
4. torch.optim.Adam.


In [57]:
# To use the most used Adam optimizer from PyTorch, we can simply instantiate it with:
optimizer = torch.optim.Adam(model.parameters(), lr=0.01, betas=(0.9, 0.999))
